### PR TITLE
pre-commit config

### PR DESCRIPTION
**Pull request by rachmadaniHaryono**
## Proposed changes

resolves https://github.com/bwhmather/ssort/issues/24

solution is based from this url

https://pre-commit.com/#new-hooks

this also include following change/feature

- fix typo on readme
- precommit config
- example to use pre commit config
- github workflow to run pre commit
- tox env for pre-commit

added yaml ini file is formatted using 

https://github.com/macisamuele/language-formatters-pre-commit-hooks

rst is checked using 

https://github.com/Lucas-C/pre-commit-hooks-markup

github pre commit workflow is based from 

https://github.com/pre-commit/action

isort and black pre commit config is added and modified from following urls

- https://pycqa.github.io/isort/docs/configuration/pre-commit.html
- https://black.readthedocs.io/en/stable/integrations/source_version_control.html?highlight=pre%20commit#version-control-integration

after merging this pr, the default `rev` value have to be changed to next version (`0.10.1` or `0.11.0`)

to test the pr add following hook to your pre commit config

```yaml
- repo: https://github.com/rachmadaniHaryono/ssort
  rev: 545dc67985605706f8c26d5b596e7373d4ce9cf1
  hooks:
  - id: ssort
```
adjust rev value pr status

## Types of changes

_Put an `x` in the boxes that apply_

-  Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

- I have read the CONTRIBUTING doc
- Lint and unit tests pass locally with my changes
-  I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- Any dependent changes have been merged and published in downstream modules

## Further comments

this pr is modified from 

https://github.com/appium/appium/blob/master/.github/PULL_REQUEST_TEMPLATE.md

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -94,12 +94,34 @@ To check that a file is correctly sorted use the `--check` flag.
...
 If ``ssort`` needs to make changes to a `black <https://black.readthedocs.io/en/stable/>`_ conformant file, the result will not necessarily be `black <https://black.readthedocs.io/en/stable/>`_ conformant.
 The result of running `black <https://black.readthedocs.io/en/stable/>`_ on an ``ssort`` conformant file will always be ``ssort`` conformant.
-We recommend that you reformat using `isort <https://pycqa.github.io/isort/>`_ and `black <https://black.readthedocs.io/en/stable/>`_ immediately _after_ running ``ssort``.
+We recommend that you reformat using `isort <https://pycqa.github.io/isort/>`_ and `black <https://black.readthedocs.io/en/stable/>`_ immediately after running ``ssort``.
 
 .. code:: bash
 
     $ ssort src/ tests/ setup.py; isort src/ tests/ setup.py; black src/ tests/ setup.py
 
+Pre-commit config example:
```
Comment (bwhmather) [inline README.rst:103]: I think this needs a bit more detail.  At the very least, a link to the pre-commit hooks manager package documentation, and some indication that this is optional.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    exclude: examples/.*.py
+- repo: https://github.com/psf/black
+  rev: 22.1.0
+  hooks:
+  - id: black
+    exclude: examples/.*.py
+- repo: https://github.com/Lucas-C/pre-commit-hooks-markup
+  rev: v1.0.1
+  hooks:
+  - id: rst-linter
```
Comment (bwhmather) [inline .pre-commit-config.yaml:31]: All checks in here need to be enforced by github actions and, ideally, tox to prevent commits by people using different tools from containing lots of unrelated formatting changes.  Do you know if rst-linter, pretty-format-yaml, and the four from the pre-commit org have standalone equivalents?

File: README.rst [hunk 3]
```diff
@@ -94,12 +94,34 @@ To check that a file is correctly sorted use the `--check` flag.
...
+.. code:: yaml
+
+
+  repos:
+  # ...
+  - repo: https://github.com/bwhmather/ssort
+    rev: 0.10.0
+    hooks:
+    - id: ssort
+  - epo: https://github.com/pycqa/isort
```
Comment (bwhmather) [inline README.rst:114]: Should this be `repo`?

File: README.rst (hunk 1 above)
Comment (rachmadaniHaryono) [reply README.rst:103]: like this?

```rst
or you can setup `pre-commit <https://pre-commit.com/index.html>` hook, for example:```

File: .pre-commit-config.yaml (hunk 2 above)
Comment (rachmadaniHaryono) [reply .pre-commit-config.yaml:31]: > - All checks in here need to be enforced by github actions and,
> - ideally, tox to prevent commits by people using different tools from containing lots of unrelated formatting changes

for github action there is this https://github.com/pre-commit/action, which mean i need to create new github workflow yaml

second point : i just realize that you use tox to test the repo. i can convert those hooks to use tox using pre commit local hook https://pre-commit.com/#repository-local-hooks

>  Do you know if rst-linter, pretty-format-yaml, and the four from the pre-commit org have standalone equivalents?

can you explain more about this point?

but as note only rst-linter and language-formatters-pre-commit-hooks not pypi but both can be installed from git repo

four hooks from pre-commit are from pre commit itself

**Comment by rachmadaniHaryono** [comment]
added 3 hooks, 2 local using tox and 1 for ini file

i also add new env for tox to run pre-commit

that pre-commit will skip pylint and pyflakes hooks to avoid double check when running full tox

i also add github workflow for pre-commit but i can't test it until it is merged

pr post is updated to match recent commits

>> All checks in here need to be enforced by github actions and,
>> ideally, tox to prevent commits by people using different tools from containing lots of unrelated formatting changes

> second point : i just realize that you use tox to test the repo. 
> i can convert those hooks to use tox using pre commit local hook pre-commit.com/#repository-local-hooks

i said that but black and isort hook are different than what is configured on tox.ini

i also can't convert test env from tox to pre commit because it raise error when developer don't have required python

**Comment by bwhmather** [comment]
Let's split out the `.pre-commit-hooks.yaml` change, as that's what I think is most urgent.

Regarding the `.pre-commit-config.yaml`, I think it's safer if it just always calls into tox and not the other way around.  As we have the same tests running in CI, `tox --skip-missing-intererpreters` should be safe.   I'd like to get that rst-linter thing into tox and CI, as it looks particulary useful.

**Comment by rachmadaniHaryono** [comment]
how about this part?

>>I think this needs a bit more detail. At the very least, a link to the pre-commit hooks manager package documentation, and some indication that this is optional.

> like this?

> ```rst
> or you can setup `pre-commit <https://pre-commit.com/index.html>` hook, for example:
> ```

**Comment by bwhmather** [comment]
Perhaps:

> You can also setup ssort to run automatically before commit by creating setting up `pre-commit <https://pre-commit.com/index.html>`, and registering ssort in your `.pre-commit-config.yaml`.

**Comment by bwhmather** [comment]
Thank you for not directly copying and pasting my garbled attempt at english.  I've had a long day...

**Comment by rachmadaniHaryono** [comment]
for this part

    `.pre-commit-config.yaml`
    
there may be difference between single and double backtick string

https://stackoverflow.com/questions/22256995/restructuredtext-in-sphinx-and-docstrings-single-vs-double-back-quotes-or-back

https://docutils.sourceforge.io/docs/user/rst/quickref.html#inline-markup

i just choose single backtick without any reason

**Comment by bwhmather** [comment]
Looks ready to go.  Would you be able to clean up the history using `git rebase --interactive`?  Good resource is https://git-rebase.io/, if you aren't familiar with how it works.  Should be relatively easy here as you've been disciplined about keeping your commits atomic.

**Comment by rachmadaniHaryono** [comment]
i assume repo owner can set it up when pr merged?

https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github#squashing-your-merge-commits

**Comment by bwhmather** [comment]
Don't want to squash everything, just make it look like you jumped straight to the answer.  Merge the typo fixes into the commits that introduced the typo, merge commits to add and remove files, that sort of thing.  Squash erases history and is bad.  Rebase gives you an artificially perfect history, and makes you look superhuman.

**Review by bwhmather — APPROVED** [review]
Fantastic, thank you.

**Comment by rachmadaniHaryono** [comment]
as note

> after merging this pr, the default rev value have to be changed to next version (0.10.1 or 0.11.0)

i will also try the hook after pr merged

**Comment by bwhmather** [comment]
Are you able to request running the checks?  I can't find a way to trigger them from this end.

**Comment by bwhmather** [comment]
I'll do a release straight after.

**Comment by rachmadaniHaryono** [comment]
i think it is  caused by github action that i add before, i will try something

**Comment by bwhmather** [comment]
You can possibly do `rev: feature/pre-commit` with URL set to your repo?

**Comment by rachmadaniHaryono** [comment]
> You can possibly do rev: feature/pre-commit with URL set to your repo?

no

https://pre-commit.com/index.html#pre-commit-configyaml---repos

> the revision or tag to clone at. new in 1.7.0: previously sha

**Comment by rachmadaniHaryono** [comment]
there is different solution here https://stackoverflow.com/questions/52200096/github-pull-request-waiting-for-status-to-be-reported

can you try close the pr and reopen it again https://stackoverflow.com/a/67986789

**Comment by bwhmather** [comment]
Force merged as very unlikely to cause test failures.  Have pushed to pypi as v0.10.1.

**Comment by rachmadaniHaryono** [comment]
>    after merging this pr, the default rev value have to be changed to next version (0.10.1 or 0.11.0)

i try it and it should be updated to 

```yaml
- repo: https://github.com/bwhmather/ssort
  rev: v0.10.1
  hooks:
  - id: ssort
```

**Comment by bwhmather** [comment]
It realistically needs to be part of the release process.  I'll do a 0.10.2 release and fix the docs.  No idea how to automate it.  If I haven't scared you away then a `.pre-commit-config.yaml` PR would be the next step.

**Comment by rachmadaniHaryono** [comment]
> It realistically needs to be part of the release process 

no it just have to be minimum supported version, which is `v0.10.1` not `0.10.0` like from this pr

note also that `v` character

maybe mention `pre commit autoupdate` so user can get latest ssort version?
